File: scripts/compare_rttm.py
# /// script
# requires-python = ">=3.10"
# dependencies = []
# ///
"""Compare two RTTM files and report differences."""

import sys
from collections import defaultdict


def parse_rttm(path):
    segments = []
    with open(path) as f:
        for line in f:
            parts = line.strip().split()
            if not parts or parts[0] != "SPEAKER":
                continue
            start = float(parts[3])
            dur = float(parts[4])
            speaker = parts[7]
            segments.append((start, dur, speaker))
    return segments


def total_speech(segments):
    return sum(dur for _, dur, _ in segments)


def speaker_stats(segments):
    by_speaker = defaultdict(float)
    for _, dur, spk in segments:
        by_speaker[spk] += dur
    return dict(by_speaker)


def timeline_overlap(segs_a, segs_b):
    """Compute how much of segs_a's speech time is covered by segs_b (any speaker)"""
    # build sorted intervals for b
    intervals_b = sorted([(s, s + d) for s, d, _ in segs_b])
    if not intervals_b:
        return 0.0

    # merge overlapping intervals in b
    merged = [intervals_b[0]]
    for start, end in intervals_b[1:]:
        if start <= merged[-1][1]:
            merged[-1] = (merged[-1][0], max(merged[-1][1], end))
        else:
            merged.append((start, end))

    covered = 0.0
    for s, d, _ in segs_a:
        seg_start, seg_end = s, s + d
        for b_start, b_end in merged:
            if b_start >= seg_end:
                break
            if b_end <= seg_start:
                continue
            overlap_start = max(seg_start, b_start)
            overlap_end = min(seg_end, b_end)
            covered += max(0, overlap_end - overlap_start)
    return covered


def main():
    if len(sys.argv) != 3:
        print("Usage: compare_rttm.py <rust.rttm> <python.rttm>", file=sys.stderr)
        sys.exit(1)

    rust_segs = parse_rttm(sys.argv[1])
    py_segs = parse_rttm(sys.argv[2])

    rust_total = total_speech(rust_segs)
    py_total = total_speech(py_segs)
    rust_speakers = speaker_stats(rust_segs)
    py_speakers = speaker_stats(py_segs)

    print(f"{'':30s} {'Rust':>10s} {'Python':>10s}")
    print(f"{'─' * 52}")
    print(f"{'Segments':30s} {len(rust_segs):10d} {len(py_segs):10d}")
    print(f"{'Speakers':30s} {len(rust_speakers):10d} {len(py_speakers):10d}")
    print(f"{'Total speech (s)':30s} {rust_total:10.1f} {py_total:10.1f}")
    print()

    print("Per-speaker duration (s):")
    print(f"  Rust:   {rust_speakers}")
    print(f"  Python: {py_speakers}")
    print()

    # speech overlap analysis
    if rust_total > 0 and py_total > 0:
        rust_covered = timeline_overlap(rust_segs, py_segs)
        py_covered = timeline_overlap(py_segs, rust_segs)
        print(f"{'Rust speech covered by Python':30s} {rust_covered / rust_total * 100:9.1f}%")
        print(f"{'Python speech covered by Rust':30s} {py_covered / py_total * 100:9.1f}%")

        # time range
        all_segs = rust_segs + py_segs
        audio_end = max(s + d for s, d, _ in all_segs)
        print(f"{'Audio span (s)':30s} {audio_end:10.1f}")


if __name__ == "__main__":
    main()

File: src/bin/compare_rttm.rs
use std::collections::HashMap;
use std::fs;

struct Segment {
    start: f64,
    duration: f64,
    speaker: String,
}

fn parse_rttm(path: &str) -> Vec<Segment> {
    let content = fs::read_to_string(path).unwrap_or_else(|e| {
        eprintln!("failed to read {path}: {e}");
        std::process::exit(1);
    });

    content
        .lines()
        .filter_map(|line| {
            let parts: Vec<&str> = line.split_whitespace().collect();
            if parts.first() != Some(&"SPEAKER") || parts.len() < 8 {
                return None;
            }
            Some(Segment {
                start: parts[3].parse().ok()?,
                duration: parts[4].parse().ok()?,
                speaker: parts[7].to_string(),
            })
        })
        .collect()
}

fn total_speech(segments: &[Segment]) -> f64 {
    segments.iter().map(|s| s.duration).sum()
}

fn speaker_durations(segments: &[Segment]) -> HashMap<&str, f64> {
    let mut map = HashMap::new();
    for seg in segments {
        *map.entry(seg.speaker.as_str()).or_insert(0.0) += seg.duration;
    }
    map
}

fn format_speaker_durations(durations: &HashMap<&str, f64>) -> String {
    let mut pairs: Vec<_> = durations.iter().collect();
    pairs.sort_by_key(|(k, _)| k.to_string());
    let inner: Vec<String> = pairs.iter().map(|(k, v)| format!("{k}: {v:.1}s")).collect();
    format!("{{{}}}", inner.join(", "))
}

/// Compute how much of segs_a's speech time is covered by segs_b (any speaker)
fn timeline_overlap(segs_a: &[Segment], segs_b: &[Segment]) -> f64 {
    let mut intervals: Vec<(f64, f64)> = segs_b
        .iter()
        .map(|s| (s.start, s.start + s.duration))
        .collect();
    intervals.sort_by(|a, b| a.0.partial_cmp(&b.0).unwrap());

    if intervals.is_empty() {
        return 0.0;
    }

    // merge overlapping intervals
    let mut merged = vec![intervals[0]];
    for &(start, end) in &intervals[1..] {
        let last = merged.last_mut().unwrap();
        if start <= last.1 {
            last.1 = last.1.max(end);
        } else {
            merged.push((start, end));
        }
    }

    let mut covered = 0.0;
    for seg in segs_a {
        let seg_start = seg.start;
        let seg_end = seg.start + seg.duration;
        for &(b_start, b_end) in &merged {
            if b_start >= seg_end {
                break;
            }
            if b_end <= seg_start {
                continue;
            }
            covered += seg_end.min(b_end) - seg_start.max(b_start);
        }
    }
    covered
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() != 3 {
        eprintln!("Usage: compare_rttm <rust.rttm> <python.rttm>");
        std::process::exit(1);
    }

    let rust_segs = parse_rttm(&args[1]);
    let py_segs = parse_rttm(&args[2]);

    let rust_total = total_speech(&rust_segs);
    let py_total = total_speech(&py_segs);
    let rust_speakers = speaker_durations(&rust_segs);
    let py_speakers = speaker_durations(&py_segs);

    println!("{:30} {:>10} {:>10}", "", "Rust", "Python");
    println!("{}", "─".repeat(52));
    println!(
        "{:30} {:10} {:10}",
        "Segments",
        rust_segs.len(),
        py_segs.len()
    );
    println!(
        "{:30} {:10} {:10}",
        "Speakers",
        rust_speakers.len(),
        py_speakers.len()
    );
    println!(
        "{:30} {:>10.1} {:>10.1}",
        "Total speech (s)", rust_total, py_total
    );
    println!();

    println!("Per-speaker duration:");
    println!("  Rust:   {}", format_speaker_durations(&rust_speakers));
    println!("  Python: {}", format_speaker_durations(&py_speakers));
    println!();

    if rust_total > 0.0 && py_total > 0.0 {
        let rust_covered = timeline_overlap(&rust_segs, &py_segs);
        let py_covered = timeline_overlap(&py_segs, &rust_segs);
        println!(
            "{:30} {:>9.1}%",
            "Rust speech covered by Python",
            rust_covered / rust_total * 100.0
        );
        println!(
            "{:30} {:>9.1}%",
            "Python speech covered by Rust",
            py_covered / py_total * 100.0
        );

        let audio_end = rust_segs
            .iter()
            .chain(py_segs.iter())
            .map(|s| s.start + s.duration)
            .fold(0.0_f64, f64::max);
        println!("{:30} {:>10.1}", "Audio span (s)", audio_end);
    }
}
